scrapers/csci_topics_scraper/main.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+
+# Python standard library
+import asyncio
+from operator import itemgetter
+import os
+import re
+import json
+import sys
+
+# External dependnecies
+import aiohttp
+import bs4
+import requests
+
+BeautifulSoup = lambda data: bs4.BeautifulSoup(data, features="lxml")
+
+
+async def get_topics_txts():
+    global session
+    url = "https://www.cs.rpi.edu/~goldsd/docs/"
+    async with session.get(url) as request:
+        soup = BeautifulSoup(await request.text())
+        links = soup.find("table").findAll("a")
+        out = []
+        for link in links:
+            if "topics-courses" in link["href"]:
+                out.append(link["href"])
+        return out
+
+
+async def scrape_all_txts(links):
+    list_of_terms = {}
+    for link in links:
+        await scrape_txt(link, list_of_terms)
+    return list_of_terms
+
+
+def goldy_term_to_sis_term(term):
+    term = term.lower()
+    if term[0] == "s":
+        postfix = "01"
+    elif term[0] == "u":
+        postfix = "05"
+    elif term[0] == "f":
+        postfix = "09"
+    else:
+        # placeholder
+        postfix = "00"
+    return "20" + term[1:] + postfix
+
+
+async def scrape_txt(filename, list_of_terms):
+    global session
+    url = f"https://www.cs.rpi.edu/~goldsd/docs/{filename}"
+    async with session.get(url) as request:
+        data = re.split(r"\n======*", (await request.text()))
+        for term_data in data:
+            courses = [
+                line
+                for line in re.split("\n\n", term_data.replace("\r", ""))
+                if line.strip()
+            ]
+            term = courses[0].replace("[", "").split(" ")[0]
```

Comment:
I'm just going to assume that this works.

Reply:
It's kind of horrible, but basically Goldschmidt uses an inconsistent number of newlines between blocks of information in his txt files, and he also uses Windows which results in `\r` showing up. So what this code does is first get rid of all `\r` (regex doesn't work properly if I don't do that; this took me like an hour to debug), and it removes entries from `courses` that are just empty strings, which break the rest of the logic.

Finally the reason why I remove left brackets at the end is because sometimes Goldschmidt uses them in the heading of the file (https://www.cs.rpi.edu/~goldsd/docs/s23-topics-courses.txt) and sometimes he doesn't (https://www.cs.rpi.edu/~goldsd/docs/S21-topics-courses.txt).

Reply:
Added a comment explaining what is happening, for everyone's sake.